File: Daily_Challenge/20230113_LC_1061.cpp
// Written by: Jian-Yuan Lin
// Date: 2023.01.13

class Solution {
public:
    int find(int x, vector<int>& root){
        if (x != root[x]){
            root[x] = find(root[x], root);
        }
        return root[x];
    }

    void do_union(int x, int y, vector<int>& root){
        int rx = find(x, root), ry = find(y, root);
        if (rx != ry){
            if (rx < ry){
                root[ry] = rx;
            }
            else{
                root[rx] = ry;
            }
        }
    }
    string smallestEquivalentString(string s1, string s2, string baseStr) {
        vector<int> root(26, 0);
        for (int i = 0; i < 26; i++){
            root[i] = i;
        }
        for (int i = 0; i < s1.size(); i++){
            char c1 = s1[i], c2 = s2[i];
            do_union(int(c1) - int('a'), int(c2) - int('a'), root);
        }
        string chars = "abcdefghijklmnopqrstuvwxyz";
        string ans = "";
        for (int i = 0; i < baseStr.size(); i++){
            ans += chars[find(int(baseStr[i]) - int('a'), root)];
        }
        return ans;
    }
};

File: Daily_Challenge/20230113_LC_1061.py
# Written by: Jian-Yuan Lin
# Date: 2023.01.13

class Solution:
    def smallestEquivalentString(self, s1: str, s2: str, baseStr: str) -> str:
        root = [i for i in range(26)]

        def find(x):
            if x != root[x]:
                root[x] = find(root[x])
            return root[x]
        
        def union(x, y):
            rx, ry = find(x), find(y)
            if rx != ry:
                if rx < ry:
                    root[ry] = rx
                else:
                    root[rx] = ry
        
        for i in range(len(s1)):
            c1, c2 = s1[i], s2[i]
            if c1 != c2:
                union(ord(c1) - ord('a'), ord(c2) - ord('a'))
        
        chars = "abcdefghijklmnopqrstuvwxyz"
        ans = ""
        for c in baseStr:
            ans += chars[find(ord(c) - ord('a'))]
        return ans